operator/buffer/disk.go

```diff
@@ -42,26 +49,348 @@ func NewDiskBufferConfig() *DiskBufferConfig {
+		return err
+	}
+
+	// Assert for sanity; This should never occur
+	if curEnd != int64(d.end) {
+		return fmt.Errorf("the current EOF (%d) was not equal to the expected EOF (%d)", curEnd, d.end)
+	}
+
+	_, err = d.f.Write(bufBytes)
+	if err != nil {
+		return err
+	}
+
+	d.end += int64(len(bufBytes))
+
+	// Signal to a potentially waiting reader that there is an entry to read
+	d.readReady.Signal()
+
 	return nil
 }
 
 // Read reads from the buffer.
-// Read will block until the there are MaxChunkSize entries or we have block as long as MachChunkDelay.
-func (m *DiskBuffer) Read(ctx context.Context) ([]*entry.Entry, error) {
-	return nil, nil
+// Read will block until the there are maxChunkSize entries or the duration maxChunkDelay has passed.
+func (d *DiskBuffer) Read(ctx context.Context) ([]*entry.Entry, error) {
+	entries := make([]*entry.Entry, 0)
+	timer := time.NewTimer(d.maxChunkDelay)
+	defer timer.Stop()
+
+	d.mux.Lock()
+	defer d.mux.Unlock()
+
+	if d.closed {
+		return nil, ErrBufferClosed
+	}
+
+	for len(entries) < int(d.maxChunkSize) {
+		for d.end <= int64(d.metadata.StartOffset) {
+			// No entries to read
+			timerDone := waitCondWithCtxAndTimer(ctx, d.readReady, timer)
+			if timerDone {
+				return entries, nil
+			}
+
+			select {
+			case <-ctx.Done():
+				return nil, fmt.Errorf("got context done: %w", ctx.Err())
+			default:
+			}
+
+			if d.closed {
+				return nil, ErrBufferClosed
+			}
+		}
+
+		// Seek to the start position of the file
+		_, err := d.f.Seek(int64(d.metadata.StartOffset), io.SeekStart)
+		if err != nil {
+			return nil, err
+		}
+
+		var entry entry.Entry
+		dec := json.NewDecoder(d.f)
+
+		err = dec.Decode(&entry)
+		if err != nil {
+			return nil, err
+		}
+
+		entries = append(entries, &entry)
+
+		decoderOffset := dec.InputOffset()
+
+		// Update start pointer to current position
+		d.metadata.StartOffset += uint64(decoderOffset) + 1
+		err = d.metadata.Sync(d.f)
+		if err != nil {
+			return nil, err
+		}
+
+		// Signal to the writers that they may be able to write again, since we freed up some space
+		d.writeReady.Broadcast()
+	}
+
+	return entries, nil
 }
 
-// Close runs cleanup code for buffer
-func (m *DiskBuffer) Close() ([]*entry.Entry, error) {
+// Close runs cleanup code for buffer.
+func (d *DiskBuffer) Close() ([]*entry.Entry, error) {
+	d.mux.Lock()
+	defer d.mux.Unlock()
+
+	if d.closed {
+		return nil, nil
+	}
+
+	err := d.f.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	d.closed = true
+
+	// Wake up readers/writers so they don't block while the buffer is closed
+	d.readReady.Broadcast()
+	d.writeReady.Broadcast()
+
 	return nil, nil
 }
+
+// compactBufferSize is the size of the internal buffer used when reading/writing during compaction.
+var compactBufferSize int64 = 1 << 12 // 4KiB
+
+// compact compacts the file by moving all data backwards to the start of file, then truncating the file
+// to the new buffer length.
+func (d *DiskBuffer) compact() error {
```

Comment:
I need to rewrite this to be more resilient to restarts; Just realized it could be a problem here.